### PR TITLE
sys/compat/vtimer: get rid of

### LINKED ISSUE #4601: sys: some modules are using obsolete vtimer
Let's convert them, then get rid of vtimer.

- [ ] sys/posix/pthread
- [ ] sys/net/routing/nhdp

### PR DESCRIPTION
**Pull request by kYc0o**
Second part of #4660 

fixes #4601

### CODE REVIEW

File: Makefile.dep [hunk 1]
```diff
@@ -8,7 +8,7 @@ endif
 
 ifneq (,$(filter nhdp,$(USEMODULE)))
   USEMODULE += conn_udp
-  USEMODULE += vtimer
+  USEMODULE += xtimer
```
Comment (cgundogan) [inline Makefile.dep]: this line should be #5039

**Comment by gebart** [comment]
tests/pthread_condition_variable has problems building

**Comment by kYc0o** [comment]
Yes there still are some vtimers... I'll try to fix it

**Comment by gebart** [comment]
I don't understand either CI's failures... RIOT CI failed because of a segfault in pyterm (!?), Travis fails because the compiler doesn't find C++11 headers when building for _some_ of the ARM boards...

**Comment by kaspar030** [comment]
@gebart I think the unittest segfaults, and then pexpect dies over it in a very eloquent way.

**Comment by gebart** [comment]
weird that the unittests ran OK in Travis: https://travis-ci.org/RIOT-OS/RIOT/jobs/115750231

**Comment by kaspar030** [comment]
I think RIOT CI is slow enough to show another problem: With this PR, if I leave the unittests binary on native running (after the completed tests), it segfaults after about 15 secs. In master, they don't.

```
$ /home/kaspar/src/riot/tests/unittests/bin/native/unittests.elf
RIOT native interrupts/signals initialized.
LED_GREEN_OFF
LED_RED_ON
RIOT native board initialized.
RIOT native hardware initialization complete.

main(): This is RIOT! (Version: 2016.03-devel-1009-g16df9-booze-test)
..........................................................................................................................................................................
..........................................................................................................................................................................
..........................................................................................................................................................................
.............................................................................................................................................
OK (651 tests)
Segmentation fault (core dumped)
[kaspar@booze unittests (test)]$
```

**Comment by kaspar030** [comment]
I managed to get a backtrace:
```
(gdb) r
Starting program: /home/kaspar/src/riot/tests/unittests/bin/native/unittests.elf
RIOT native interrupts/signals initialized.
LED_GREEN_OFF
LED_RED_ON
RIOT native board initialized.
RIOT native hardware initialization complete.

main(): This is RIOT! (Version: 2016.03-devel-1009-g16df9-booze-test)
...........................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................
OK (651 tests)

Program received signal SIGSEGV, Segmentation fault.
0x00000000 in ?? ()
(gdb) bt
#0  0x00000000 in ?? ()
#1  0x0804bf59 in _shoot (timer=0x81031bc <ncache+60>) at xtimer_core.c:146
#2  0x0804c4ef in _timer_callback () at xtimer_core.c:474
#3  0x0804bf3e in _periph_timer_callback (arg=0x0, chan=0) at xtimer_core.c:141
#4  0x080a53fa in native_isr_timer () at timer.c:75
#5  0x0804ad4a in native_irq_handler () at irq_cpu.c:255
#6  0xf7dbe2db in makecontext () from /usr/lib32/libc.so.6
#7  0x00000000 in ?? ()
(gdb)
```

**Comment by kYc0o** [comment]
So that shows a problem with xtimer? Since I replaced vtimer by xtimer in the pthread test. Maybe the test is not adapted for xtimer?

**Comment by kaspar030** [comment]
It definately looks like a stray timer. But why in "ncache+60"?

**Comment by gebart** [comment]
@kaspar030 stack overflow?

**Comment by kaspar030** [comment]
No, it's definately a stray timer. enabling xtimer's debug shows a timer set with offset=15000000, and the crash happens when the timer triggers.

**Comment by kaspar030** [comment]
This is where the timer gets set:

```
(gdb) r
Starting program: /home/kaspar/src/riot/tests/unittests/bin/native/unittests.elf
RIOT native interrupts/signals initialized.
LED_GREEN_OFF
LED_RED_ON
RIOT native board initialized.
RIOT native hardware initialization complete.

main(): This is RIOT! (Version: 2016.03-devel-1009-g16df9-booze-test)
_lltimer_set(): setting 4294967295
............................................................................................................................................................................................................................................................................................................................................................................................................................................................................
Breakpoint 1, xtimer_set (timer=0x81031bc <ncache+60>, offset=15000000) at xtimer_core.c:119
119         DEBUG("timer_set(): offset=%" PRIu32 " now=%" PRIu32 " (%" PRIu32 ")\n", offset, xtimer_now(), _lltimer_now());
(gdb) bt
#0  xtimer_set (timer=0x81031bc <ncache+60>, offset=15000000) at xtimer_core.c:119
#1  0x0804ba86 in xtimer_set_msg (timer=0x81031bc <ncache+60>, offset=15000000, msg=0x81031d0 <ncache+80>, target_pid=0) at xtimer.c:124
#2  0x0809b762 in gnrc_ndp_internal_reset_nbr_sol_timer (nc_entry=0x8103180 <ncache>, delay=15000000, type=535, pid=0)
    at ../../../../../include/net/gnrc/ndp/internal.h:207
#3  0x0809bc8c in gnrc_ipv6_nc_still_reachable (ipv6_addr=0x80f49c7 <main_stack+11623>) at gnrc_ipv6_nc.c:254
#4  0x0806db10 in test_ipv6_nc_still_reachable__success () at tests-ipv6_nc.c:389
#5  0x08097de4 in TestCase_run (self=0x80f4a38 <main_stack+11736>, result=0x8103120 <result_>) at TestCase.c:58
#6  0x0809781c in TestCaller_run (self=0x80cbfe8 <ipv6_nc_tests>, result=0x8103120 <result_>) at TestCaller.c:54
#7  0x080979c0 in TestRunner_runTest (test=0x80cbfe8 <ipv6_nc_tests>) at TestRunner.c:99
#8  0x0806dec3 in tests_ipv6_nc () at tests-ipv6_nc.c:479
#9  0x0804c7c7 in main () at main.c:36
(gdb)
```

**Comment by cgundogan** [comment]
@kYc0o could you kick out commit 17d668a75929a57ffd30625d71c03dca7a292abf and then rebase to current master?

**Comment by cgundogan** [comment]
@kaspar030 can you give more insight (backtrace) on the error that murdock is facing here?

**Comment by kYc0o** [comment]
Any Murdock expert around?

**Comment by cgundogan** [comment]
@kaspar030 I cannot reproduce the error from murdock.
I use `make -C ./tests/unittests -B clean all test BOARD=native".
The tests run pretty quickly and then make terminates successfully. How did you reproduce that?

**Comment by cgundogan** [comment]
nevermind, `BUILD_IN_DOCKER=1 BUILDTEST_MCU_GROUP=x86 ./dist/tools/travis-scripts/build_and_test.sh` did the trick

**Comment by cgundogan** [comment]
hm still can't reproduce. travis seems fine though. Maybe we should skip murdocks opinion this time, since the error doesn't seem to be related to this PR.

**Comment by kaspar030** [comment]
> the error doesn't seem to be related to this PR.

Are we sure about that? (I'll restart the build...)

**Comment by cgundogan** [comment]
hmm, the offending change could be
```
-#if defined(MODULE_GNRC_IPV6_NETIF) && defined(MODULE_VTIMER) && defined(MODULE_GNRC_IPV6)
 +#if defined(MODULE_GNRC_IPV6_NETIF) && defined(MODULE_XTIMER) && defined(MODULE_GNRC_IPV6)
```
`MODULE_VTIMER` is most probably not set in current builds other than this PR and therefore this code path never executed. So it may be related to this PR :/

**Comment by cgundogan** [comment]
the above diff is for `sys/net/gnrc/network_layer/ipv6/nc/gnrc_ipv6_nc.c`

**Comment by kaspar030** [comment]
That seems likely, and no other builds crash at that point...

**Comment by kaspar030** [comment]
Could you run a shell in docker, then try to run the unittest there in gdb?

**Comment by cgundogan** [comment]
I did the following:
```
> docker run -i -t -v $(pwd):/data/riotbuild riot/riotbuild /bin/bash
> apt-get update
> apt-get install gdb
> make -C ./tests/unittests -B clean all-debug debug BOARD=native
> gdb> run
...
...
...
main(): This is RIOT! (Version: 2016.03-devel-1196-g85f37-5a58ba63d5ca-kYc0o-rid_of_vtimer)
.............................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................
OK (653 tests)
(here it hands forever)
```
Now I wait for about 5 minutes and nothing happens.

**Comment by cgundogan** [comment]
even if I start the binary like you did in your output above it doesn't fail for me:
```
root@5a58ba63d5ca:/data/riotbuild/tests/unittests/bin/native# ./unittests.elf
RIOT native interrupts/signals initialized.
LED_RED_OFF
LED_GREEN_ON
RIOT native board initialized.
RIOT native hardware initialization complete.

main(): This is RIOT! (Version: 2016.03-devel-1196-g85f37-5a58ba63d5ca-kYc0o-rid_of_vtimer)
.............................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................
OK (653 tests)
```

I keep it open for minutes and nothing happens.

**Comment by cgundogan** [comment]
I am just wondering and this may be a long shot, but can it be that the broken version of this PR (before rebasing to the fixed master) is cached in ccache?

**Comment by cgundogan** [comment]
@RIOT-OS/maintainers anybody care to look into that? I have no clues how to tackle this issue.

**Comment by cgundogan** [comment]
Does this PR needs a rebase or will Murdoch pick up the changes by itself?

**Comment by kaspar030** [comment]
Murdock always rebases to master.

**Comment by OlegHahm** [comment]
```
Program received signal SIGSEGV, Segmentation fault.
0x080a140e in gnrc_ipv6_nc_still_reachable (ipv6_addr=0x810088c ) at gnrc_ipv6_nc.c:270
270	        gnrc_ndp_internal_reset_nbr_sol_timer(entry, iface->reach_time,
#0  0x080a140e in gnrc_ipv6_nc_still_reachable (ipv6_addr=0x810088c ) at gnrc_ipv6_nc.c:270
#1  0x08072243 in test_ipv6_nc_still_reachable__success () at tests-ipv6_nc.c:396
#2  0x0809cb1d in TestCase_run (self=0x81008f8 , result=0x810efe8 ) at TestCase.c:58
#3  0x0809ca86 in TestCaller_run (self=0x80d65f8 , result=0x810efe8 ) at TestCaller.c:54
#4  0x0809d008 in TestRunner_runTest (test=0x80d65f8 ) at TestRunner.c:99
#5  0x0807265b in tests_ipv6_nc () at tests-ipv6_nc.c:486
#6  0x0804c962 in main () at main.c:36
```
Looing at the code it seems that `gnrc_ipv6_netif_t *iface = gnrc_ipv6_netif_get(entry->iface);` returns a NULL pointer here for some reason.

**Comment by cgundogan** [comment]
>Looing at the code it seems that gnrc_ipv6_netif_t *iface = gnrc_ipv6_netif_get(entry->iface); returns a NULL pointer here for some reason.

Can you explain how you deduced that?

**Comment by cgundogan** [comment]
@kYc0o can you rebase your branch on #5169 so that we can verify this fixes everything?

**Comment by kYc0o** [comment]
done

**Comment by authmillenon** [comment]
Something went wrong.. Anyways: https://github.com/RIOT-OS/RIOT/pull/5169 was merged so you can rebase to master again ^^

**Comment by kYc0o** [comment]
sorry, my bad, I'm still a young padawan on this git stuff...

**Comment by cgundogan** [comment]
fear not .. your git-fu will improve quickly!

**Comment by kYc0o** [comment]
rebased, squashed and fixed. Thanks a lot @cgundogan

**Comment by authmillenon** [comment]
Murdock is happy. Was there an ACK?

**Comment by cgundogan** [comment]
yes, but I would prefer to also wait on travis for this one as this PR had different outcomes for both CIs before

**Comment by cgundogan** [comment]
Both CIs agree. Finally we get rid of all remnants of the `vtimer`. Thanks @kYc0o !

GO

**Comment by kaspar030** [comment]
Nice one!

**Comment by OlegHahm** [comment]
Cool!

**Pull request merged**